Add new Python API `SBCommandInterpreter::GetTranscript()`

royitaqi opened the pull request:
# Motivation

Currently, the user can already get the "transcript" (for "what is the transcript", see `CommandInterpreter::SaveTranscript`). However, the only way to obtain the transcript data as a user is to first destroy the debugger, then read the save directory. Note that destroy-callbacks cannot be used, because 1\ transcript data is private to the command interpreter (see `CommandInterpreter.h`), and 2\ the writing of the transcript is *after* the invocation of destory-callbacks (see `Debugger::Destroy`).

So basically, there is no way to obtain the transcript:
* during the lifetime of a debugger (including the destroy-callbacks, which often performs logging tasks, where the transcript can be useful)
* without relying on external storage

In theory, there are other ways for user to obtain transcript data during a debugger's life cycle:
* Use Python API and intercept commands and results.
* Use CLI and record console input/output.

However, such ways rely on the client's setup and are not supported natively by LLDB.


# Proposal

Add a new Python API `SBCommandInterpreter::GetTranscript()`.

Goals:
* It can be called at any time during the debugger's life cycle, including in destroy-callbacks.
* It returns data in-memory.

Structured data:
* To make data processing easier, the return type is `SBStructuredData`. See comments in code for how the data is organized.
* In the future, `SaveTranscript` can be updated to write different formats using such data (e.g. JSON). This is probably accompanied by a new setting (e.g. `interpreter.save-session-format`).

# Alternatives

The return type can also be `std::vector<std::pair<std::string, SBCommandReturnObject>>`. This will make implementation easier, without having to translate it to `SBStructuredData`. On the other hand, `SBStructuredData` can convert to JSON easily, so it's more convenient for user to process.

# Privacy

Both user commands and output/error in the transcript can contain privacy data. However, as mentioned, the transcript is already available to the user. The addition of the new API doesn't increase the level of risk. In fact, it _lowers_ the risk of privacy data being leaked later on, by avoiding writing such data to external storage.

Once the user (or their code) gets the transcript, it will be their responsibility to make sure that any required privacy policies are guaranteed.

# Tests

```
bin/llvm-lit -sv ../external/llvm-project/lldb/test/API/python_api/interpreter/TestCommandInterpreterAPI.py
```

```
bin/llvm-lit -sv ../external/llvm-project/lldb/test/API/commands/session/save/TestSessionSave.py
```

royitaqi committed:
Add SBCommandInterpreter::GetTranscript()

clayborg reviewed (changes requested):
You will need to hook this up so it actually works, and then add a test case to verify this works as expected.

medismailben commented:
I think you forgot to include the implementation.

royitaqi committed:
Implement the new API

royitaqi committed:
Add unittest

medismailben commented on lldb/source/API/SBCommandInterpreter.cpp:
```diff
@@ -571,6 +571,17 @@ SBStructuredData SBCommandInterpreter::GetStatistics() {
+  LLDB_INSTRUMENT_VA(this);
+
+  SBStructuredData data;
+  if (!IsValid())
+    return data;
+
+  data.m_impl_up->SetObjectSP(m_opaque_ptr->GetTranscript());
+  return data;
```
```suggestion
  if (IsValid())
    data.m_impl_up->SetObjectSP(m_opaque_ptr->GetTranscript());
  return data;
```

medismailben commented on lldb/include/lldb/Utility/StructuredData.h:
```diff
@@ -290,6 +290,9 @@ class StructuredData {
 
     void GetDescription(lldb_private::Stream &s) const override;
 
+    static ArraySP SplitString(llvm::StringRef s, char separator, int maxSplit,
+                               bool keepEmpty);
```
Could you document this method ? It's not clear what `maxSplit` & `keepEmpty` do

royitaqi replied:
Moved. New document is in the header file, explaining all params and return.

medismailben commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -2044,6 +2050,15 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
+  // Add output and error to the transcript item after splitting lines. In the
+  // future, other aspects of the command (e.g. perf) can be added, too.
+  transcript_item->AddItem(
+      "output", StructuredData::Array::SplitString(result.GetOutputData(), '\n',
+                                                   -1, false));
+  transcript_item->AddItem(
+      "error", StructuredData::Array::SplitString(result.GetErrorData(), '\n',
+                                                  -1, false));
```
Looks like you could set the `separator`, `maxSplit` & `keepEmpty` default value to what you have here.

medismailben commented on lldb/source/Utility/StructuredData.cpp:
```diff
@@ -289,3 +292,35 @@ void StructuredData::Null::GetDescription(lldb_private::Stream &s) const {
   s.Printf("%p", m_object);
 }
+
+/// This is the same implementation as `StringRef::split`. Not depending on
+/// `StringRef::split` because it will involve a temporary `SmallVectorImpl`.
```
This comment should be in the header file

royitaqi replied:
Moved and expanded.

medismailben commented on lldb/source/Utility/StructuredData.cpp:
```diff
@@ -289,3 +292,35 @@ void StructuredData::Null::GetDescription(lldb_private::Stream &s) const {
+                                                           bool keepEmpty) {
+  auto array_sp = std::make_shared<StructuredData::Array>();
+
+  // Count down from MaxSplit. When MaxSplit is -1, this will just split
+  // "forever". This doesn't support splitting more than 2^31 times
+  // intentionally; if we ever want that we can make MaxSplit a 64-bit integer
+  // but that seems unlikely to be useful.
```
nit: match variable casing
```suggestion
  // Count down from `maxSplit`. When `maxSplit` is -1, this will just split
  // "forever". This doesn't support splitting more than 2^31 times
  // intentionally; if we ever want that we can make `maxSplit` a 64-bit integer
  // but that seems unlikely to be useful.
```

medismailben commented on lldb/source/Utility/StructuredData.cpp:
```diff
@@ -289,3 +292,35 @@ void StructuredData::Null::GetDescription(lldb_private::Stream &s) const {
+
+/// This is the same implementation as `StringRef::split`. Not depending on
+/// `StringRef::split` because it will involve a temporary `SmallVectorImpl`.
+StructuredData::ArraySP StructuredData::Array::SplitString(llvm::StringRef s,
```
Although I see the use of this method in the patch, I think it should either land in a different patch or this patch should at least add some unit tests for it.

royitaqi replied:
Thanks for the prompt review. I appreciate it.

> or this patch should at least add some unit tests for it.
For sure. Let me add (and also other improvements that you have pointed out).

clayborg replied:
This whole function is a small wrapper around StringRef::split(). I am not sure this specialized method belongs in StructuredData (and I would like to not split the strings for output and error if possible).

royitaqi committed:
Add more test asserts and some touch ups

royitaqi committed:
Apply suggestions from code review

Co-authored-by: Med Ismail Bennani <ismail@bennani.ma>

royitaqi committed:
Move and add document for Array::SplitString

royitaqi committed:
Add unit test for Array::SplitString

royitaqi reviewed (commented):
Updated according to @medismailben 's suggestions.

royitaqi committed:
Fix format

royitaqi committed:
Improve Python API test reliability

bulbazord commented on lldb/include/lldb/API/SBCommandInterpreter.h:
```diff
@@ -318,6 +318,12 @@ class SBCommandInterpreter {
+  /// Returns a list of handled commands, output and error. Each element in
+  /// the list is a dictionary with three keys: "command" (string), "output"
+  /// (list of strings) and optionally "error" (list of strings). Each string
+  /// in "output" and "error" is a line (without EOL characteres).
```
typo: `characteres` -> `characters`

bulbazord commented on lldb/include/lldb/Interpreter/CommandInterpreter.h:
```diff
@@ -766,6 +768,12 @@ class CommandInterpreter : public Broadcaster,
+  /// the list is a dictionary with three keys: "command" (string), "output"
+  /// (list of strings) and optionally "error" (list of strings). Each string
+  /// in "output" and "error" is a line (without EOL characters).
+  StructuredData::ArraySP m_transcript_structured;
```
nit: No need to put `structured` in the variable name.

royitaqi replied:
There is another variable right above this one, called `m_transcript_stream`. So I added the `structured` suffix to emphasize the difference.

bulbazord commented on lldb/include/lldb/Utility/StructuredData.h:
```diff
@@ -290,6 +290,36 @@ class StructuredData {
+    /// Note:
+    /// * This is almost the same API and implementation as `StringRef::split`.
+    /// * Not depending on `StringRef::split` because it will involve a
+    ///   temporary `SmallVectorImpl`.
```
It sounds like this is justifying duplication for efficiency reasons. Have you actually observed that there is an actual gain in duplicating this functionality here? If you're doing this preemptively because it sounds like it should be faster, I would say measure or don't do it at all.

royitaqi replied:
Updated the implementation to call `StringRef::split` with a `SmallVector`, then move the substrings into the to-be-returned array.

bulbazord commented on lldb/include/lldb/Utility/StructuredData.h:
```diff
@@ -366,10 +396,10 @@ class StructuredData {
+        : Object(lldb::eStructuredDataTypeString), m_value(s) {}
 
-    void SetValue(llvm::StringRef S) { m_value = std::string(S); }
+    void SetValue(llvm::StringRef s) { m_value = std::string(s); }
```
The changes to adjust the name of the variable are unnecessary and make it harder to navigate commit history. Please revert it.

royitaqi replied:
This improves readability/consistency within the file. Agreed about navigation of commit history. Will revert.

royitaqi committed:
Update implementation of Array::SplitString, and other smaller changes

bulbazord commented on lldb/source/Utility/StructuredData.cpp:
```diff
@@ -289,3 +290,19 @@ void StructuredData::Null::GetDescription(lldb_private::Stream &s) const {
+  // Copy the substrings from the small vector into the output array.
+  auto array_sp = std::make_shared<StructuredData::Array>();
+  for (auto substring : small_vec) {
+    array_sp->AddStringItem(std::move(substring));
```
nit: no need for the move, StringRefs are pretty cheap to copy

bulbazord commented on lldb/test/API/python_api/interpreter/TestCommandInterpreterAPI.py:
```diff
@@ -85,3 +86,84 @@ def test_command_output(self):
+        # The transcript will contain a bunch of commands that are run
+        # automatically. We only want to validate for the ones that are
+        # listed above, hence trimming to the last parts.
+        transcript = transcript[-total_number_of_commands:]
```
Instead of trimming, why not verify that the transcript contains exactly the number we care about? Or is there a reason there would be more? Perhaps from some setup code?

royitaqi replied:
*There is a lot more*. I am not sure entirely sure if it's from the set up code in the above. It seems they contain alias or custom command setup. I don't have a completely list. If you want I can print you one here in comment.

The point is that the unit test shouldn't care about any extra command that was ran that is irrelevant to this test. So trim makes sense to me.

bulbazord replied:
I think that's fine. Thanks for checking.

royitaqi replied:
Hi @bulbazord ,

> Or is there a reason there would be more? Perhaps from some setup code?
I have verified that that's indeed the case.

See the [full transcript](https://pastebin.com/Gxj1Lhid). I believe the commands that are before "version" are coming from [this general setup code](https://github.com/llvm/llvm-project/blob/main/lldb/packages/Python/lldbsuite/test/lldbtest.py#L744-L767)

So I think trimming make sense here, in order to isolate this test case from the setup code, which may be changed in the future. Let me add this to the comment so that it's clear to future readers.

bulbazord reviewed (commented):
Looking better. Thanks for sticking with us through the review! :)

royitaqi committed:
Update comment, more reliable test, minor fix

royitaqi commented:
I believe I have addressed all your comments, @bulbazord .

I appreciate your comments so far - they all make sense and are great comments to help me improve this PR. Thank you for spending time on this.

clayborg commented on lldb/include/lldb/Interpreter/CommandInterpreter.h:
```diff
@@ -766,6 +768,12 @@ class CommandInterpreter : public Broadcaster,
 
   StreamString m_transcript_stream;
+
+  /// Contains a list of handled commands, output and error. Each element in
+  /// the list is a dictionary with three keys: "command" (string), "output"
+  /// (list of strings) and optionally "error" (list of strings). Each string
+  /// in "output" and "error" is a line (without EOL characters).
```
If we have a transcript already, can we start storing the time it took the run each command and also add this to the JSON as a "time_sec" (to indicate time in seconds) as a float value?

royitaqi replied:
Yes, we definitely can. I plan to do that as a separate PR as a follow-up.

I want to land this PR sooner rather than later, in order to unblock my other work in Meta (to log this transcript into the `lldb_usage` table).

clayborg commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -2044,6 +2052,15 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
+                                                   -1, false));
+  transcript_item->AddItem(
+      "error", StructuredData::Array::SplitString(result.GetErrorData(), '\n',
+                                                  -1, false));
```
Can we time the "cmd_obj->Execute()" calls and add this as a key/value pair?

royitaqi replied:
See my rely to your other comment about I plan to do it as a separate, follow-up PR.

jimingham commented:
Sorry to come in a little late on this, but I think we need a setting to turn this off as well.  If you aren't planning to use the transcript, you shouldn't have to pay the cost for it.

royitaqi commented:
Hi @jimingham ,

> Sorry to come in a little late on this

No worries. Thanks for chiming in.

> I think we need a setting to turn this off as well. If you aren't planning to use the transcript, you shouldn't have to pay the cost for it.

Could you elaborate about why this is necessary? If we really need this, can I add the setting in a follow-up PR?

--

FWIW, to help make the conversation faster (reduce rounds of communication), I will write down my gut feeling below. Please kindly correct me if I'm wrong.

**First**, why I think the added cost is negligible:
1. The baseline is that there is already a `m_transcript_stream`, which doesn't have any settings to turn off. This PR just doubles that cost (plus some string split operations). It doesn't add huge cost on top of this baseline.
2. Maintaining transcript is a per-command operation, and commands are relatively infrequently invoked. Here I assume the common use cases run tens or at most hundreds of commands in a debug session. The above cost times 10x ~ 100x isn't much.
3. The cost of the commands themselves (and the process) are much higher than that of the transcript. E.g. resolving breakpoints and printing variables all need to do actual work, rather than just recording info.

**Secondly**, if we are going to add settings, there needs to be product design considerations _based on user usage/feedback_, so we probably need more time to gather data points after shipping this feature.
1. Prematurely add settings (which we cannot remove, I assume) can lead us down to a more difficult road.
2. I think it will be the norm that we record transcripts, like bash/zsh records command histories, so by default this should be on. In fact it is already, by `m_transcript_stream`.
3. Besides just a binary on/off, settings can also allow more granular control. E.g. verbosity; only log transcript for a subset of commands; limit the memory footprint by using a cyclic buffer; etc. If we add these settings, they will cause the on/off setting to be redundant and needs to be removed.
4. (As mentioned in "Secondly") The above are all possibilities. We need user usage/feedback to guide the design of these settings. So I think we should ship this PR first then gather data points.

jimingham commented:
> On May 7, 2024, at 11:28 AM, royitaqi ***@***.***> wrote:
> 
> 
> Hi @jimingham <https://github.com/jimingham> ,
> 
> Sorry to come in a little late on this
> 
> No worries. Thanks for chiming in.
> 
> I think we need a setting to turn this off as well. If you aren't planning to use the transcript, you shouldn't have to pay the cost for it.
> 
> Could you elaborate about why this is necessary? If we really need this, can I add the setting in a follow-up PR?
> 
> --
> 
> FWIW, to help make the conversation faster (reduce rounds of communication), I will write down my gut feeling below. Please kindly correct me if I'm wrong.
> 
> First, why I think the added cost is negligible:
> 
> The baseline is that there is already a m_transcript_stream, which doesn't have any settings to turn off. This PR just doubles that cost (plus some string split operations). It doesn't add huge cost on top of this baseline.
That's equally an argument for allowing me to turn that off as well - if I never plan to use `session save` then it's silly to take up memory for that either.  

> Maintaining transcript is a per-command operation, and commands are relatively infrequently invoked. Here I assume the common use cases run tens or at most hundreds of commands in a debug session. The above cost times 10x ~ 100x isn't much.
People have automated scripts that run debug sessions, and you are also storing output text so this can get fairly big.  Plus if you start adding time stamps we're stopping to get the system clock on every command, it gets more expensive.  For people running lldb on big beefy machines, this is unlikely to matter for, but people also run lldb on low memory devices, where resources are tighter.


> The cost of the commands themselves (and the process) are much higher than that of the transcript. E.g. resolving breakpoints and printing variables all need to do actual work, rather than just recording info.
I don't think "unrelated thing X is bigger" is a a useful data point when considering whether it's good to do Y.


> Secondly, if we are going to add settings, there needs to be product design considerations based on user usage/feedback, so we probably need more time to gather data points after shipping this feature.
> 
> Prematurely add settings (which we cannot remove, I assume) can lead us down to a more difficult road.

Not sure why a setting to turn off recording transcripts would ever be controversial.

> I think it will be the norm that we record transcripts, like bash/zsh records command histories, so by default this should be on. In fact it is already, by m_transcript_stream.
I guess.  I must admit I've used lldb since it existed and have never once used `session save`.  After all, my Terminal has a nice big buffer...  For my workflow that's all wasted memory.


> Besides just a binary on/off, settings can also allow more granular control. E.g. verbosity; only log transcript for a subset of commands; limit the memory footprint by using a cyclic buffer; etc. If we add these settings, they will cause the on/off setting to be redundant and needs to be removed.
I don't see the force of this argument at all.  Having a control "whether to do X at all" and then another set "if I were to do X, how should I do it" is a pretty standard approach.  As a matter of fact, that's a better design since then I can disable the feature temporarily without having to mess up the settings I want when I do do it.

> (As mentioned in "Secondly") The above are all possibilities. We need user usage/feedback to guide the design of these settings. So I think we should ship this PR first then gather data points.
I really don't think this is controversial, and I'm not sure how you would gather data points. If it helps to get started, however, I vote there should be a way to turn this off.

But if you are running out of time for this feature and really need to get it off your plate, deferring the ability to turn it off to a subsequent PR is okay.

Jim

> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/90703#issuecomment-2099053680>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW5NYBNLLRY6ETNKPN3ZBEMMFAVCNFSM6AAAAABHBPBMWOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJZGA2TGNRYGA>.
> You are receiving this because you were mentioned.
>

royitaqi commented:
Hi @jimingham,

Thank you so much for the detailed explanation about the context and your thoughts. I appreciate it.

Good to know about the automated scripts and that you personally never used the existing transcript. It makes me feel that there are distinctively different use cases, and for the ones where the user know they won't need the transcript, they should be able to turn them off (and yes the text output can be large if someone opt to print e.g. statistics dump).

> As a matter of fact, that's a better design since then I can disable the feature temporarily without having to mess up the settings I want when I do do it.

This is also an interesting design aspect that I never thought about. This is good learning.

--

Overall I feel what you said makes sense. I will open a follow-up PR specifically for the settings, where we can discuss more (e.g. default on or off; should the setting turn off the existing transcript; should we add a `interpreter.save-session-format`; etc).

jimingham commented:
Sounds good to me.

Jim


> On May 7, 2024, at 4:52 PM, royitaqi ***@***.***> wrote:
> 
> 
> Hi @jimingham <https://github.com/jimingham>,
> 
> Thank you so much for the detailed explanation about the context and your thoughts. I appreciate it.
> 
> Good to know about the automated scripts and that you personally never used the existing transcript. It makes me feel that there are distinctively different use cases, and for the ones where the user know they won't need the transcript, they should be able to turn them off (and yes the text output can be large if someone opt to print e.g. statistics dump).
> 
> As a matter of fact, that's a better design since then I can disable the feature temporarily without having to mess up the settings I want when I do do it.
> 
> This is also an interesting design aspect that I never thought about. This is good learning.
> 
> --
> 
> Overall I feel what you said makes sense. I will open a follow-up PR specifically for the settings, where we can discuss more (e.g. default on or off; should the setting turn off the existing transcript; should we add a interpreter.save-session-format; etc).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/90703#issuecomment-2099490330>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW3UEZVXIJWIB64IBITZBFSNTAVCNFSM6AAAAABHBPBMWOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJZGQ4TAMZTGA>.
> You are receiving this because you were mentioned.
>

clayborg commented on lldb/test/API/python_api/interpreter/TestCommandInterpreterAPI.py:
```diff
@@ -85,3 +86,91 @@ def test_command_output(self):
+        ci.HandleCommand("an-unknown-command", res)
+        ci.HandleCommand("breakpoint set -f main.c -l %d" % self.line, res)
+        ci.HandleCommand("r", res)
+        ci.HandleCommand("p a", res)
```
Add an example that gets some JSON as outoput like "statistics dump". We need to test that a command can output anything, including things that contain JSON itself and we need to make sure the output string or error string that contains JSON doesn't mess up the structured data (the strings for output need to be sanitized to desensitize any JSON reserved characters (like '{' and '"', etc).

royitaqi replied:
Added the "statistics dump" test command. Validated that 1\ the output is valid JSON (through `json.loads`) and 2\ the output contains expected fields, like "commands" and "modules".

clayborg reviewed (changes requested):
Needs a few changes. See inline comments.

royitaqi committed:
Remove Array::SplitString + add test for JSON output

royitaqi committed:
Clean-ups

royitaqi commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -2044,6 +2052,15 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
+                                                   -1, false));
+  transcript_item->AddItem(
+      "error", StructuredData::Array::SplitString(result.GetErrorData(), '\n',
+                                                  -1, false));
```
Removed the split, and the utility function. Now `StructuredData.h/.cpp` are unchanged.

royitaqi commented on lldb/include/lldb/Utility/StructuredData.h:
```diff
@@ -290,6 +290,31 @@ class StructuredData {
+    ///   string.
+    static ArraySP SplitString(llvm::StringRef s, char separator,
+                               int maxSplit = -1, bool keepEmpty = true);
+
```
Same as the above.
> Removed the split, and the utility function. Now `StructuredData.h/.cpp` are unchanged.

royitaqi committed:
Use assertIn in python test

jeffreytan81 commented on lldb/include/lldb/API/SBCommandInterpreter.h:
```diff
@@ -247,13 +247,13 @@ class SBCommandInterpreter {
-  /// Returns whether an interrupt flag was raised either by the SBDebugger - 
+  /// Returns whether an interrupt flag was raised either by the SBDebugger -
   /// when the function is not running on the RunCommandInterpreter thread, or
   /// by SBCommandInterpreter::InterruptCommand if it is.  If your code is doing
-  /// interruptible work, check this API periodically, and interrupt if it 
+  /// interruptible work, check this API periodically, and interrupt if it
   /// returns true.
   bool WasInterrupted() const;
-  
+
```
Please undo these changes. 
In VSCode, you can select your modified code and format it instead of formatting the whole file.

royitaqi replied:
I was wondering how to only format the lines I touched. Thanks for the advice.

jeffreytan81 commented on lldb/include/lldb/Interpreter/CommandInterpreter.h:
```diff
@@ -576,7 +577,7 @@ class CommandInterpreter : public Broadcaster,
 
   bool GetRepeatPreviousCommand() const;
-  
+
```
Similar, undo the unnecessary change

jeffreytan81 commented on lldb/source/API/SBCommandInterpreter.cpp:
```diff
@@ -150,7 +150,7 @@ bool SBCommandInterpreter::WasInterrupted() const {
 bool SBCommandInterpreter::InterruptCommand() {
   LLDB_INSTRUMENT_VA(this);
-  
+
```
Similar, undo the unnecessary change

jeffreytan81 commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -135,7 +136,8 @@ CommandInterpreter::CommandInterpreter(Debugger &debugger,
       m_truncation_warning(eNoOmission), m_max_depth_warning(eNoOmission),
-      m_command_source_depth(0) {
+      m_command_source_depth(0),
+      m_transcript(std::make_shared<StructuredData::Array>()) {
```
This looks weird. Simply use default constructor is fine: `m_transcript()`

royitaqi replied:
Doesn't the default ctor create an empty shared pointer? Here what we want is to create the object, thus `make_shared`.

royitaqi replied:
This is out-dated now. In the latest commit, the field is now a `StructuredData::Array`. The initializer has been removed.

jeffreytan81 commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -1891,6 +1893,12 @@ bool CommandInterpreter::HandleCommand(const char *command_line,
 
+  // The same `transcript_item` will be used below to add output and error of
+  // the command.
+  auto transcript_item = std::make_shared<StructuredData::Dictionary>();
```
You are not sharing it. Use `std::make_unique`

royitaqi replied:
Two reasons for using the `shared_ptr`:
1. `m_transcript` holds a `shared_ptr` anyways. The `m_transcript->AddItem(transcript_item)` call below this line also requires a `shared_ptr`.
2. Many lines below this line, the `transcript_item` is used again in order to add the command's output and error.

If we use `unique_ptr` instead, then the code will look like:
```
  auto transcript_item = std::make_unique<StructuredData::Dictionary>();
  transcript_item->AddStringItem("command", command_line);
  m_transcript->AddItem(std::move(transcript_item));

  ...

  auto last_transcript_item = m_transcript.last();
  last_transcript_item->AddStringItem("output", result.GetOutputData());
  last_transcript_item->AddStringItem("error", result.GetErrorData());
```

I feel the last part is weird (i.e. first creating `transcript_item` and move into `m_transcript`, then getting it again from `m_transcript`). I feel a `shared_ptr` is more readable and performant (since `m_transcript` holds a `shared_ptr` anyways, rather than adding then getting).

WDYT?

jeffreytan81 reviewed (commented):
General looks good to me. I do agree that we should hide this behind a setting to avoid memory leak in long running lldb sessions.

clayborg commented on lldb/include/lldb/Interpreter/CommandInterpreter.h:
```diff
@@ -766,6 +768,12 @@ class CommandInterpreter : public Broadcaster,
+  /// the list is a dictionary with three keys: "command" (string), "output"
+  /// (list of strings) and optionally "error" (list of strings). Each string
+  /// in "output" and "error" is a line (without EOL characters).
+  StructuredData::ArraySP m_transcript;
```
This shouldn't be a shared pointer, we don't want to hand out references to this structured data across the API boundary because it can be accessed by the caller and also be updated by new commands coming in. This should be a `StructuredData::Array` object and we should copy it when we hand it out across the API boundary.

clayborg replied:
Or we can leave this as a shared pointer and when  people call GetTranscript() on this object we give away our copy and start a new one.

clayborg replied:
Not sure if this would make sense to the user though...

royitaqi replied:
Improve the loop as discussed (I think).

clayborg commented on lldb/source/API/SBCommandInterpreter.cpp:
```diff
@@ -571,6 +571,15 @@ SBStructuredData SBCommandInterpreter::GetStatistics() {
+
+  SBStructuredData data;
+  if (IsValid())
+    data.m_impl_up->SetObjectSP(m_opaque_ptr->GetTranscript());
```
Here don't want to hand out a shared pointer to the internal transcript. If we do and the caller starts accessing it, we can crash if new commands are added to the structured data. If we to not copy things, one thing we could do it to grab the transript shared pointer from the command interpreter and have it create a new one.

royitaqi replied:
Made a (deep) copy in the latest commit. Added test to verify that the user's copy isn't modified when more commands are run in the debugger.

royitaqi committed:
Add a setting; add elapsed seconds; make a copy of the returned StructuredData::Array; add tests

royitaqi committed:
Fix format

royitaqi committed:
Attempt to revert unnecessary format changes

royitaqi committed:
Fix TestSessionSave.py; update a comment

clayborg commented on lldb/include/lldb/Interpreter/CommandInterpreter.h:
```diff
@@ -766,6 +771,14 @@ class CommandInterpreter : public Broadcaster,
 
   StreamString m_transcript_stream;
+
+  /// Contains a list of handled commands and their details. Each element in
```
We should mention the new setting that is required for this to become populated.

clayborg commented on lldb/include/lldb/Interpreter/CommandInterpreter.h:
```diff
@@ -766,6 +771,14 @@ class CommandInterpreter : public Broadcaster,
   CommandUsageMap m_command_usages;
 
   StreamString m_transcript_stream;
```
We should mention in a comment that this will only be filled in if the setting is true.

clayborg reviewed (commented):
Just a few comments to fixup as mentioned and this is good to go.

royitaqi committed:
Add comments the new setting
